priam/src/main/java/com/netflix/priam/backup/BackupVerification.java

```diff
@@ -33,39 +37,84 @@
 public class BackupVerification {
 
     private static final Logger logger = LoggerFactory.getLogger(BackupVerification.class);
-    private final IMetaProxy metaProxy;
+    private final IMetaProxy metaV1Proxy;
+    private final IMetaProxy metaV2Proxy;
+    private final IBackupStatusMgr backupStatusMgr;
     private final Provider<AbstractBackupPath> abstractBackupPathProvider;
 
     @Inject
     BackupVerification(
-            @Named("v1") IMetaProxy metaProxy,
+            @Named("v1") IMetaProxy metaV1Proxy,
+            @Named("v2") IMetaProxy metaV2Proxy,
+            IBackupStatusMgr backupStatusMgr,
             Provider<AbstractBackupPath> abstractBackupPathProvider) {
-        this.metaProxy = metaProxy;
+        this.metaV1Proxy = metaV1Proxy;
+        this.metaV2Proxy = metaV2Proxy;
+        this.backupStatusMgr = backupStatusMgr;
         this.abstractBackupPathProvider = abstractBackupPathProvider;
     }
 
-    public Optional<BackupMetadata> getLatestBackupMetaData(List<BackupMetadata> metadata) {
-        return metadata.stream()
-                .filter(backupMetadata -> backupMetadata != null)
-                .filter(backupMetadata -> backupMetadata.getStatus() == Status.FINISHED)
-                .sorted(Comparator.comparing(BackupMetadata::getStart).reversed())
-                .findFirst();
+    private IMetaProxy getMetaProxy(int backupVersion) {
+        switch (backupVersion) {
+            case SnapshotBackup.BACKUP_VERSION:
+                return metaV1Proxy;
+            case SnapshotMetaService.BACKUP_VERSION:
+                return metaV2Proxy;
+        }
+
+        return null;
     }
 
-    public Optional<BackupVerificationResult> verifyBackup(List<BackupMetadata> metadata) {
-        if (metadata == null || metadata.isEmpty()) return Optional.empty();
+    public Optional<BackupVerificationResult> verifyBackup(
+            int backupVersion, boolean force, DateRange dateRange)
+            throws UnsupportedTypeException, IllegalArgumentException {
+        IMetaProxy metaProxy = getMetaProxy(backupVersion);
+        if (metaProxy == null) {
+            throw new UnsupportedTypeException(
+                    "BackupVersion type: " + backupVersion + " is not supported");
+        }
 
-        Optional<BackupMetadata> latestBackupMetaData = getLatestBackupMetaData(metadata);
+        if (dateRange == null) {
+            throw new IllegalArgumentException("dateRange provided is null");
+        }
 
-        if (!latestBackupMetaData.isPresent()) {
-            logger.error("No backup found which finished during the time provided.");
-            return Optional.empty();
+        List<BackupMetadata> metadata =
+                backupStatusMgr.getLatestBackupMetadata(backupVersion, dateRange);
+        if (metadata == null || metadata.isEmpty()) return Optional.empty();
+        for (BackupMetadata backupMetadata : metadata) {
```

Comment:
So for backup 2.0, do we get the guarantee that old snapshots will eventually get all their data because we enumerate all snapshots and try to upload them regardless of when they happened (resumable backup)?

Reply:
Yes, you are correct. We try to upload all the files to achieve resume. We try to upload at the start of Priam as well to ensure that if backup 2.0 cron is set far apart, we still try to upload those files and not wait until next run.